File: 2025/2/gift_shop.py
INPUT_FILE = f"{__file__.replace('.py', '_input.txt')}"
INPUT_FILE_SMALL = f"{__file__.replace('.py', '_input_small.txt')}"

def get_input(file):
    with open(file) as f:
        return [tuple(int(split) for split in s.split('-')) for s in f.read().replace('\n', '').split(',')]

def part1(values):
    invalid_id_sum = 0
    for start, stop in values:
        for val in range(start, stop + 1):
            str_val = str(val)
            str_len = len(str_val) // 2
            if str_val[:str_len] == str_val[str_len:]:
                invalid_id_sum += val
    print(f"First part: {invalid_id_sum}")

def part2(values):
    invalid_id_sum = 0
    for start, stop in values:
        for val in range(start, stop + 1):
            str_val = str(val)
            stop = 1
            # print("--- val:", val)
            while stop <= len(str_val) / 2:
                if str_val.count(str_val[:stop]) == len(str_val) / stop:
                    invalid_id_sum += val
                    break
                stop += 1
    print(f"Second part: {invalid_id_sum}")

def main():
    values = get_input(INPUT_FILE)
    # values = get_input(INPUT_FILE_SMALL)
    part1(values)  # First part: 18700015741
    part2(values)  # Second part: 20077272987

if __name__ == "__main__":
    main()

# Elegant solution found on Reddit (credit to 4HbQ):
# It uses regex to find digit group
# for lo, hi in findall(r"(\d+)-(\d+)", *open("in.txt")):
#     for i in range(int(lo), int(hi) + 1):
#         if match(r"^(\d+)\1$", str(i)):
#             a += i
#         if match(r"^(\d+)\1+$", str(i)):
#             b += i

File: rust/2025/day02/gift_shop.rs

fn get_input(small: bool) -> Vec<(u64, u64)> {
    let file = aoc_utils::get_input_filename!(small);
    std::fs::read_to_string(&file)
        .expect("Failed to read file")
        .replace("\n", "")
        .split(",")
        .map(|line| {
            let parts = line.split('-').collect::<Vec<&str>>();
            (
                parts[0].parse::<u64>().unwrap(),
                parts[1].parse::<u64>().unwrap(),
            )
        })
        .collect()
}

fn part1(input: &Vec<(u64, u64)>) -> u64 {
    let mut invalid_id_sum = 0;
    for (start, stop) in input {
        for id in *start..=*stop {
            let id_str = id.to_string();
            let id_len = id_str.len() / 2;
            if id_str[..id_len] == id_str[id_len..] {
                invalid_id_sum += id;
            }
        }
    }
    invalid_id_sum
}

fn part2(input: &Vec<(u64, u64)>) -> u64 {
    let mut invalid_id_sum = 0;
    for (start, stop) in input {
        for id in *start..=*stop {
            let id_str = id.to_string();
            let id_len = id_str.len();
            for pattern_len in 1..=id_len / 2 {
                if id_len % pattern_len == 0 {
                    let pattern = &id_str[..pattern_len];
                    let repeat_count = id_len / pattern_len;
                    if pattern.repeat(repeat_count) == id_str {
                        invalid_id_sum += id;
                        break;
                    }
                }
            }
        }
    }
    invalid_id_sum
}

fn main() {
    let input_small = get_input(true);
    println!("First part small: {}", part1(&input_small));
    println!("Second part small: {}", part2(&input_small));

    println!("--------------------------------");

    let input = get_input(false);
    println!("First part: {}", part1(&input));
    println!("Second part: {}", part2(&input));
}

// Results:
// First part: 18700015741
//Second part: 20077272987
